haystack/nodes/file_converter/parsr.py

```diff
@@ -0,0 +1,237 @@
+        row_idx_start = 0
+        caption = ""
+        table_list = [[""] * len(element["content"][0]["content"])
+                      for _ in range(len(element["content"]))]
+
+        for row_idx, row in enumerate(element["content"]):
+            for col_idx, cell in enumerate(row["content"]):
+                # Check if first row is a merged cell spanning whole table
+                # -> exclude this row and use as caption
+                if (row_idx == col_idx == 0) and (cell["colspan"] == len(table_list[0])):
+                    cell_paragraphs = [self._get_paragraph_string(par) for par in cell["content"]]
+                    cell_content = "\n\n".join(cell_paragraphs)
+                    caption = cell_content
+                    row_idx_start = 1
+                    table_list.pop(0)
+                    break
+
+                if cell["type"] == "table-cell":
+                    cell_paragraphs = [self._get_paragraph_string(par) for par in cell["content"]]
+                    cell_content = "\n\n".join(cell_paragraphs)
+                    for c in range(cell["colspan"]):
+                        for r in range(cell["rowspan"]):
+                            table_list[row_idx + r - row_idx_start][col_idx + c] = cell_content
+
+        # Get preceding and following elements of table
+        preceding_elements = []
+        following_elements = []
+        for cur_page_idx, cur_page in enumerate(all_pages[:page_idx]):
+            for cur_elem_index, elem in enumerate(cur_page["elements"]):
+                if (elem["type"] in ["paragraph", "heading"]) \
+                        and (self.remove_page_headers and "isHeader" not in elem["properties"]) \
+                        and (self.remove_page_footers and "isFooter" not in elem["properties"]):
+                    if cur_page_idx < page_idx:
+                        preceding_elements.append(elem)
+                    elif cur_page_idx == page_idx:
+                        if cur_elem_index < elem_idx:
+                            preceding_elements.append(elem)
+                        elif cur_elem_index > elem_idx:
+                            following_elements.append(elem)
+                    elif cur_page_idx > page_idx:
+                        following_elements.append(elem)
+
+        preceding_context = "\n\n".join([self._get_paragraph_string(elem)
+                                         for elem in preceding_elements]) + f"\n\n{caption}"
+        preceding_context = preceding_context.strip()
+        following_context = "\n\n".join([self._get_paragraph_string(elem)
+                                         for elem in following_elements])
+        following_context = following_context.strip()
+
+        table_meta = copy.deepcopy(meta)
+        if isinstance(table_meta, dict):
+            table_meta["preceding_context"] = preceding_context
+            table_meta["following_context"] = following_context
+        else:
```

Comment:
The `else` branch is executed if `meta` is `None` right? So `table_meta` is either a `dict` or `None`. Maybe we can change the code by first checking if `meta` is not None, then execute `deepcopy()` and set the contexts. In the else branch we set `table_meta` to a new dict?

Reply:
Done, this should look more readable now I guess.